src/result.spec.ts

```diff
@@ -49,38 +52,67 @@ describe("Result", () => {
 
         it("promise to async map to map", async () => {
             const start = AsyncOk(Promise.resolve(12));
-            const result1 = start.map((x) => {
+            const result1: AsyncOk<number> = start.map((x) => {
                 return Promise.resolve(x + 1);
             });
-            const result = result1.map((x) => x + 1);
+            const result: AsyncOk<number> = result1.map((x) => x + 1);
 
             expect(await result.value).toEqual(Some(14));
             expect((await result).value).toBe(14);
             expect(await result.error).toEqual(None);
             expect((await result).error).toBeUndefined();
         });
+
+        it("map to err should be err", () => {
+            const start = Err(new Error("Error"));
+
+            const result: Err<Error> = start.map((x: number) => x + 1);
+
+            expect(result.value).toBeUndefined();
+            expect(result.error).toEqual(new Error("Error"));
+        });
+
+        it("Result map should be Result", () => {
+            const start = 1 > 0 ? Ok(2) : Err("whatever");
+
+            const result: Result<number, string> = start.map((x) => x + 1);
+
+            expect(result.ok).toBe(true);
+            expect(result.value).toEqual(3);
+        });
+
+        it("Result map of promise should be AsyncResult", async () => {
+            const start = -1 > 0 ? Ok(2) : Err("whatever");
+
+            // Type 'AsyncOk<number> | Err<string>' is not assignable to type 'AsyncResult<number, string>'.
+            const result: AsyncResult<number, string> = start.map((x) => Promise.resolve(x + 1));
```

Comment:
@francescortiz A weird consequence of Err.map not returning promise when the callback is a promise.
It is not an issue when you await, and I think it won't have any consequnce when chaining. It is just a bit unituitive.
Maybe we could just not expose AsyncResult, so the type they would put here would be the more precise `AsyncOk<number> | Err<string>`?

Reply:
If the promise ends up rejecting we will have an `AsyncErr`, so we need to expect an `AsyncResult` to cover bor both paths:

```typescript
    const result: AsyncResult<number, string> = start.map((x) => Promise.reject("AsyncErr expected"));
```